ch8/exercise8-1/clockwall/clockwall.go

```diff
@@ -0,0 +1,54 @@
+package main
+
+import (
+	"bufio"
+	"io"
+	"log"
+	"net"
+	"os"
+	"strings"
+	"time"
+)
+
+type clock struct {
+	title string
+	host  string
+}
+
+func main() {
+	clockServers := []clock{}
+	times := os.Args[1:]
+	for _, cl := range times {
+		attrs := strings.Split(cl, "=")
+		newClock := clock{attrs[0], attrs[1]}
+		clockServers = append(clockServers, newClock)
+	}
+	for _, c := range clockServers {
+		conn, err := net.Dial("tcp", c.host) //connect to server that was created by clock
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer conn.Close() // wait until goroutine has finished
+		go c.printTimes(conn)
+	}
+
+	for {
+		time.Sleep(time.Second * 30)
+	}
+}
+
+func (c clock) printTimes(conn io.Reader) {
+	out := os.Stdout
+	scanner := bufio.NewScanner(conn) //scans the in the output from server in clock
+	for scanner.Scan() {
+		output, err := out.WriteString(c.title + ": " + scanner.Text())
```

Comment:
We discussed using channels to make sure stdout doesn't have multiple routines writing to it at the same time.  After looking at the chapter a bit more: you're welcome to do this, but you're also correct that it is beyond the material covered for this exercise.

Alternatively, you may use the `log` package to write to the console instead of `os.Stdout.Write`.  Log is concurrency safe, as seen here:

https://golang.org/src/log/log.go#L152

Mutexes and semaphores won't be covered until next chapter, but that doesn't mean you can't leverage ones that others wrote.

Reply:
I looked into mutexes. I am honestly shocked it is introduced in the next chapter instead of this one, it would have saved a lot of time trying to avoid multiple processes writing to stdout easily. I used `log` this time though.

Reply:
I can see how they would be useful though. If multiple routines are trying to change a state, only one of them can at a time.

Reply:
The reason mutexes are introduced next chapter is rather pedantic:

TGPL is written expecting readers to be familiar with most concepts from another programming language. As such, it focuses on what is unique about go.  Using goroutines & channels are a noteable part of what makes go valuable -- this approach to concurrency isn'tas easy in other languages.  Mutexes are much older, more common common and are used to allow patterns that go generally discourages.

In this particular case, it is because go is interfacing with UNIX/the stdout stream that mutexes become useful.  When dealing with "legacy" syscalls, "legacy" approaches to concurrency are appropriate. 

There is a go mantra: "communicate to share memory, don't share memory to communicate".  Mutexes general enable the latter whereas channels enable the former.  It may take months to years to learn all the various nuances and failure modes that make this mantra valuable.